### PR TITLE
New filehandler

### PR DESCRIPTION
**Pull request by PatrikLundgren**
Most of filehandler rewritten to use Qt lib and QJson. A large portion of changes are in the test files. 
Lots of code for reading and writing projects to file was made redundent by Json so all these methods were removed ( and their related structures ).

However, the structure of the filehandler hasnt changed much, the interface to other modules is virtually the same. The tests are now a lot more rigorous, i suggest looking att these tests before examining the rest of the code, possibly suggest additional tests for better coverage of edge cases. I've tested all the functionality at runtime as well.

### CODE REVIEW

File: ViAn/Filehandler/filehandler.h [hunk 1]
```diff
@@ -15,93 +15,107 @@
...
 #include "dir.h"
 #include <algorithm>
-
-
-
+#include <QJsonObject>
+#include <QFile>
+#include <QDir>
+#include <QTextStream>
+#include <QJsonDocument>
```
Comment (henpe071) [inline ViAn/Filehandler/filehandler.h]: Please seperate the different things with an empty line. At least there should be one after all the includes.

File: ViAn/Filehandler/filehandler.h [hunk 2]
```diff
@@ -15,93 +15,107 @@
...
-    //directory manipulation
-    //varying implementation
-    ID create_directory(std::string dir_path);
-    FH_ERROR delete_directory(ID id);
+//  Directory methods
+    ID create_directory(QString dir_path);
+    bool delete_directory(ID id);
 
-    //file manipulation
+//  File methods
```
Comment (henpe071) [inline ViAn/Filehandler/filehandler.h]: Grouping the functions are good but do it in the same way. First the descriptive comment then the functions and then an empty line before the next grouping. Also try not to have a lot of empty lines in one group. Some are ok if they are like subgroups.
like:
//Cars
volvo
saab

//fruits
apple
pear

File: ViAn/Filehandler/filehandler.h [hunk 3]
```diff
@@ -15,93 +15,107 @@
...
+    ID add_dir(QDir dir);
+    void add_dir(ID dir_id, QDir dir);
     ID load_project_file(std::string file_path, std::stringstream& proj_file_stream);
     void load_proj_files(std::string str);
     //add used for loading project from file
-    void add_file(ID id , std::string file_path);
+    void add_file(ID id , QString file);
 
     /**
      * @brief m_projects, m_fileMap, m_dirMap
```
Comment (henpe071) [inline ViAn/Filehandler/filehandler.h]: Change the comment to match the names of the variables.

File: ViAn/Filehandler/filehandler.h [hunk 4]
```diff
@@ -15,93 +15,107 @@
...
      */
     std::map<ID,Project*> projects;
-    std::map<ID, std::string> file_map;
-    std::map<ID, std::string> dir_map;
+    std::map<ID, QString> file_map;
+    std::map<ID, QDir> dir_map;
     /**
      * @todo implement smarter lock mechanism to avoid overhead
      * of only 1 reader/writer at a time
      * @brief dirMapLock, fileMapLock, projMapLock
```
Comment (henpe071) [inline ViAn/Filehandler/filehandler.h]: Same here change the comment to match the variable names. Also please have an empty line before the comment, between line 107/108.

File: ViAn/Filehandler/filehandler.h [hunk 5]
```diff
@@ -15,93 +15,107 @@
...
-    std::mutex proj_map_lock;// lock for handling project write/read
-    ID project_id; //counter for project ids
-    ID file_id; //counter for file ids
-    ID dir_id; //counter for directory ids
+    std::mutex dir_map_lock;    // lock for handling directory write/read
+    std::mutex file_map_lock;   // lock for handling file write/read
+    std::mutex proj_map_lock;   // lock for handling project write/read
+    ID project_id;              //counter for project ids
+    ID file_id;                 //counter for file ids
+    ID dir_id;                  //counter for directory ids
```
Comment (henpe071) [inline ViAn/Filehandler/filehandler.h]: Remove these empty lines at the end of the file

File: ViAn/Filehandler/video.cpp [hunk 6]
```diff
@@ -45,3 +26,19 @@ std::stringstream& operator<<(std::stringstream& os, const Video& vid){
 bool operator==(Video v1, Video v2){
     return v1.file_path == v2.file_path;
 }
```
Comment (henpe071) [inline ViAn/Filehandler/video.cpp]: Would like some empty lines between the functions and comments here as well.

File: ViAn/GUI/mainwindow.cpp [hunk 7]
```diff
@@ -375,8 +374,7 @@ void MainWindow::on_actionAddProject_triggered() {
  * @param input the input from the user
  * @param action the action that was triggered earlier
  */
-void MainWindow::input_switch_case(ACTION action, QString qInput) {
-    std::string input = qInput.toStdString();
+void MainWindow::input_switch_case(ACTION action, QString qInput) {
```
Comment (henpe071) [inline ViAn/GUI/mainwindow.cpp]: qInput isn't the correct case.

File: ViAn/Test/filehandlertest.cpp [hunk 8]
```diff
@@ -4,76 +4,257 @@
...
+    Project* l_proj2 = file_handler->load_project(file_handler->get_dir(proj2->dir).absoluteFilePath("TEST_PROJ2.json").toStdString());
+    Project* l_proj3 = file_handler->load_project(file_handler->get_dir(proj3->dir).absoluteFilePath("TEST_PROJ3.json").toStdString());
+
+    QVERIFY(file_handler->proj_equals(*proj1,*l_proj1));
+    QVERIFY(file_handler->proj_equals(*proj2,*l_proj2));
+    QVERIFY(file_handler->proj_equals(*proj3,*l_proj3));
+
+    QVERIFY(file_handler->delete_project(proj1->id));
+    QVERIFY(file_handler->delete_project(proj2->id));
+    QVERIFY(file_handler->delete_project(proj3->id));
```
Comment (henpe071) [inline ViAn/Test/filehandlertest.cpp]: Didn't you already test deleting projects in the previous test? You shouldn't need to do it again. Also didn't you have two save_projects, then maybe you should test both of them.

File: ViAn/Test/filehandlertest.cpp [hunk 9]
```diff
@@ -4,76 +4,257 @@
...
-    std::string v1 = "video1.txt"; // names for testfiles
-    std::string v2 = "video2.txt";
-    std::string v3 = "video3.txt";
-    ID vid1 = fh->create_file(v1, proj->files->dir);  // create testfiles
-    ID vid2 = fh->create_file(v2, proj->files->dir);
-    ID vid3 = fh->create_file(v3, proj->files->dir);
-    fh->add_video(proj, fh->get_file(vid1)); //add tesfiles as videos to project.
-    fh->add_video(proj, fh->get_file(vid2));
-    fh->add_video(proj, fh->get_file(vid3));
+void FileHandlerTest::project_add_remove_items(){
```
Comment (henpe071) [inline ViAn/Test/filehandlertest.cpp]: Shouldn't this function be called project_add_remove_items_test()?

missed an "if" in the comment. "..also testS IF information.."

File: ViAn/Test/filehandlertest.cpp [hunk 10]
```diff
@@ -4,76 +4,257 @@
...
+    Project* l_proj2 = file_handler->load_project(file_handler->get_dir(proj2->dir).absoluteFilePath("TEST_PROJ2.json").toStdString());
+    Project* l_proj3 = file_handler->load_project(file_handler->get_dir(proj3->dir).absoluteFilePath("TEST_PROJ3.json").toStdString());
+
+    QVERIFY(file_handler->proj_equals(*proj1,*l_proj1));
+    QVERIFY(file_handler->proj_equals(*proj2,*l_proj2));
+    QVERIFY(file_handler->proj_equals(*proj3,*l_proj3));
+
+    QVERIFY(file_handler->delete_project(proj1->id));
+    QVERIFY(file_handler->delete_project(proj2->id));
+    QVERIFY(file_handler->delete_project(proj3->id));
```
Comment (henpe071) [inline ViAn/Test/filehandlertest.cpp]: Same comment here about delete project. 
There should be a test where you add multiple videos to one project as well, to make sure multiple videos works when saving, loading and deleting projects.

File: ViAn/Test/filehandlertest.cpp [hunk 11]
```diff
@@ -4,76 +4,257 @@
...
+ */
+void FileHandlerTest::file_delete_test(){
+    QVERIFY(file_handler->delete_file(this->file_id));
+}
+
+/**
+ * @brief filehandlertest::file_create_delete_multiple
+ * Tests creating multiple files, then deletes these.
+ */
+void FileHandlerTest::file_create_delete_multiple(){
```
Comment (henpe071) [inline ViAn/Test/filehandlertest.cpp]: Both this and the next function should have test in the name

File: ViAn/Test/filehandlertest.cpp [hunk 12]
```diff
@@ -4,76 +4,257 @@
...
+    QVERIFY(file_handler->delete_file(id2));
+    QVERIFY(file_handler->delete_file(id3));
+}
+
+/**
+ * @brief FileHandlerTest::file_read_write_init
+ * Create test file and test directory.
+ */
+void FileHandlerTest::file_read_write_init()
+{
```
Comment (henpe071) [inline ViAn/Test/filehandlertest.cpp]: Move this "måsvinge" up

File: ViAn/Test/filehandlertest.cpp [hunk 13]
```diff
@@ -4,76 +4,257 @@
...
+ * Delete test file and test directory.
+ */
+void FileHandlerTest::file_read_write_cleanup()
+{
+    QVERIFY(file_handler->delete_file(this->file_id));
+    QVERIFY(file_handler->delete_directory(this->dir_id));
 }
 
+
+
```
Comment (henpe071) [inline ViAn/Test/filehandlertest.cpp]: ew

File: ViAn/Test/filehandlertest.h [hunk 14]
```diff
@@ -6,16 +6,45 @@
...
+    void file_read_write_cleanup();
+    // Project tests
+    //void project_init_tests();
+    void project_test_init();
+    void project_create_delete_test();
+    void project_save_load_test();
+    void project_add_remove_items();
+    void project_cleanup();
 
 };
```
Comment (henpe071) [inline ViAn/Test/filehandlertest.h]: Take these empty lines at the end and insert them between the different test groups instead. To make it look a little better.

File: ViAn/GUI/mainwindow.cpp (hunk 7 above)
Comment (PatrikLundgren) [inline ViAn/GUI/mainwindow.cpp]: This isn't actually code i've changed or added, odd that its showing up

File: ViAn/Test/filehandlertest.cpp (hunk 8 above)
Comment (PatrikLundgren) [inline ViAn/Test/filehandlertest.cpp]: I disagree that i shouldnt test the delets here as well. Deleting a project with allocated content is in my mind a different case than an empty project. Yes, adding another case for save_project different calls would be nice.

File: ViAn/Filehandler/filehandler.cpp [hunk 15]
```diff
@@ -2,225 +2,223 @@
...
+        root_dir = proj->dir;                        // Use present save location
+    else
+        root_dir = this->work_space;
+
+
+    proj->dir = create_directory(get_dir(root_dir).absoluteFilePath(QString::fromStdString(proj->name)));
+    add_project(proj);                          // Add project to file sytstem
+    save_project(proj);                         // Save project file
     return proj;
 }
```
Comment (PatrikLundgren) [inline ViAn/Filehandler/filehandler.cpp]: I have not! The if was a bit wrong so it didnt work. its fixed now tho. It's more flexible than it needs to be atm. As this will be implemented later (to choose directory on GUI level)

File: ViAn/GUI/mainwindow.cpp (hunk 7 above)
Comment (henpe071) [inline ViAn/GUI/mainwindow.cpp]: Yeah that it better.

File: ViAn/Test/filehandlertest.cpp (hunk 8 above)
Comment (henpe071) [inline ViAn/Test/filehandlertest.cpp]: Sure, might as well test it again I guess.

**Comment by henpe071** [comment]
The majority of the comments are fixed. The only thing left is if there should be some more tests but that isn't that important. I leave it as "changes requested" in the meantime until the tests are added or if someone else approves the pull request and wants it merged. The tests can be added later so it's not a big deal.

File: ViAn/ViAn.pro [hunk 16]
```diff
@@ -35,8 +34,13 @@ HEADERS += Test/test_video_player.h \
...
 HEADERS += Library/customdialog.h
-
-
+#
+# Boost
+#
+win32{
+    INCLUDEPATH += C:\boost\boost_1_64_0\bin
+#    LIBS += "C:\boost\boost_1_63_0\stage\" -lboost_filesystem-vc120-mt-1_63
+}
```
Comment (svidem) [inline ViAn/ViAn.pro]: Why is this line here and outcommented?
